core/src/test/java/com/datastax/oss/driver/internal/core/config/map/MapBasedDriverConfigTest.java

```diff
@@ -0,0 +1,103 @@
+import java.util.concurrent.ConcurrentHashMap;
+import org.junit.Before;
+import org.junit.Test;
+
+public class MapBasedDriverConfigTest {
+
+  private Map<String, Map<DriverOption, Object>> optionsMap;
+  private MapBasedDriverConfig config;
+
+  @Before
+  public void setup() {
+    optionsMap = new ConcurrentHashMap<>();
+    optionsMap.put(DriverExecutionProfile.DEFAULT_NAME, new ConcurrentHashMap<>());
+    config = new MapBasedDriverConfig(optionsMap);
+  }
+
+  @Test
+  public void should_load_minimal_config_with_no_profiles() {
+    addOption(DriverExecutionProfile.DEFAULT_NAME, MockOptions.INT1, 42);
+    assertThat(config).hasIntOption(MockOptions.INT1, 42);
+  }
+
+  @Test
+  public void should_inherit_option_in_profile() {
+    addOption(DriverExecutionProfile.DEFAULT_NAME, MockOptions.INT1, 42);
+    addProfile("profile1");
+    assertThat(config)
+        .hasIntOption(MockOptions.INT1, 42)
+        .hasIntOption("profile1", MockOptions.INT1, 42);
+  }
+
+  @Test
+  public void should_override_option_in_profile() {
+    addOption(DriverExecutionProfile.DEFAULT_NAME, MockOptions.INT1, 42);
+    addOption("profile1", MockOptions.INT1, 43);
+    assertThat(config)
+        .hasIntOption(MockOptions.INT1, 42)
+        .hasIntOption("profile1", MockOptions.INT1, 43);
+  }
+
+  @Test
+  public void should_create_derived_profile_with_new_option() {
+    addOption(DriverExecutionProfile.DEFAULT_NAME, MockOptions.INT1, 42);
+    DriverExecutionProfile base = config.getDefaultProfile();
+    DriverExecutionProfile derived = base.withInt(MockOptions.INT2, 43);
+
+    assertThat(base.isDefined(MockOptions.INT2)).isFalse();
+    assertThat(derived.isDefined(MockOptions.INT2)).isTrue();
+    assertThat(derived.getInt(MockOptions.INT2)).isEqualTo(43);
+  }
+
+  @Test
+  public void should_create_derived_profile_overriding_option() {
+    addOption(DriverExecutionProfile.DEFAULT_NAME, MockOptions.INT1, 42);
+    DriverExecutionProfile base = config.getDefaultProfile();
+    DriverExecutionProfile derived = base.withInt(MockOptions.INT1, 43);
+
+    assertThat(base.getInt(MockOptions.INT1)).isEqualTo(42);
+    assertThat(derived.getInt(MockOptions.INT1)).isEqualTo(43);
+  }
+
+  @Test
+  public void should_create_derived_profile_unsetting_option() {
+    addOption(DriverExecutionProfile.DEFAULT_NAME, MockOptions.INT1, 42);
+    addOption(DriverExecutionProfile.DEFAULT_NAME, MockOptions.INT2, 43);
+    DriverExecutionProfile base = config.getDefaultProfile();
+    DriverExecutionProfile derived = base.without(MockOptions.INT2);
+
+    assertThat(base.getInt(MockOptions.INT2)).isEqualTo(43);
+    assertThat(derived.isDefined(MockOptions.INT2)).isFalse();
+  }
+
+  private void addOption(String profile, DriverOption option, Object value) {
+    addProfile(profile).put(option, value);
+  }
+
+  private Map<DriverOption, Object> addProfile(String profile) {
+    return optionsMap.computeIfAbsent(profile, name -> new ConcurrentHashMap<>());
```

Comment:
> The current implementation of `MapBasedDriverConfig` makes user to think about that nested `Map<String,<Map, Object>` every time there is an interaction with this config loader.

Yes, that map is all users should think about. The intent of this loader is to bare metal, using the simplest thing that could possibly work: if you know how to use maps, you know how to use the loader.
I don't want to introduce another API layer on top of the map, because as soon as we do that we open the door for confusion (as we saw with even trivial methods like `startProfile()` and `endProfile()` in the programmatic loader).

One downside is that building nested maps is not as user-friendly as it could be, but this loader is intended for tool builders, which is a very small minority of our target audience.

As for this particular unit test, I'll admit that the code is not very elegant (particularly relying on a class-level field), I went for quick and dirty because it's only 50 or so lines of test code. Creating some sort of `NestedMapBuilder` just for that feels overkill.